---
  output: github_document
---
  
  <!-- README.md is generated from README.Rmd. Please edit that file -->
  
```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# TX2Protein - Transcript to protein

## Background

This repository contains the code used for the analysis in the manuscript *Protein isoform detection from long-read RNA-seq data*. This manuscript describes our software [NAME OF SOFTWARE], a simplified method for intergating protein coding potential of full-length transcript structures in mass spectrometry data. [NAME OF SOFTWARE] only requires a `GTF/GFF` file and the raw proteomic dataset, typically from The ProteomeXchange Consortium, and uses [`MetaMorpheus`](https://github.com/smith-chem-wisc/MetaMorpheus) for protein identifications. 

## Data

### Gene set

We got the the Early onset or syndromic epilepsy (Version 4.77) from Genomics England PanelApp. We only included Green genes, these are the genes with the highest vidence. This panel is used as a virtual panel to analyse genome or exome data in the NHS Genomic Medicine Service; the panel will routinely be applied for clinical indication 'R59 Early onset or syndromic epilepsy' but can also be used as part of the analysis for a broader clinical presentation, where relevant.

This panel is availabel through https://panelapp.genomicsengland.co.uk/panels/402/

### Long-read RNA-seq
This consists of 9 publicly available frontal cortex samples from [ENCODE](https://www.encodeproject.org/rna-seq/long-read-rna-seq/). All samples were sequenced on the PacBio Sequel II platform and processed with the ENCODE DCC deployment of the [TALON pipeline](https://github.com/ENCODE-DCC/long-read-rna-pipeline). We downloaded the TSV and GTF files from the following samples: 

|      Sample      |    Disease status   |       Tissue           |  Sex   |       Age         |
|:-----------------|:--------------------|:-----------------------|:-------|:------------------|
| [ENCSR462COR][8] | Alzheimer's disease | middle frontal area 46 | female | 90 or above years |
| [ENCSR169YNI][9] | Control             | middle frontal area 46 | female | 90 or above years |
| [ENCSR257YUB][10]| Alzheimer's disease | middle frontal area 46 | male   | 90 or above years |
| [ENCSR690QHM][11]| Alzheimer's disease | middle frontal area 46 | female | 90 or above years |
| [ENCSR316ZTD][12]| Alzheimer's disease | middle frontal area 46 | female | 90 or above years |
| [ENCSR697ASE][13]| Alzheimer's disease | middle frontal area 46 | female | 90 or above years |
| [ENCSR094NFM][14]| Control             | middle frontal area 46 | female | 90 or above years |
| [ENCSR463IDK][15]| Control             | middle frontal area 46 | female | 79                |
| [ENCSR205QMF][16]| Control             | middle frontal area 46 | female | 88                |

[8]: https://www.encodeproject.org/experiments/ENCSR462COR/
[9]: https://www.encodeproject.org/experiments/ENCSR169YNI/
[10]: https://www.encodeproject.org/experiments/ENCSR257YUB/
[11]: https://www.encodeproject.org/experiments/ENCSR690QHM/
[12]: https://www.encodeproject.org/experiments/ENCSR316ZTD/
[13]: https://www.encodeproject.org/experiments/ENCSR697ASE/
[14]: https://www.encodeproject.org/experiments/ENCSR094NFM/
[15]: https://www.encodeproject.org/experiments/ENCSR463IDK/
[16]: https://www.encodeproject.org/experiments/ENCSR205QMF/


## Analysis steps

### 1.  Get gene set
Downloading the desired panel from PanelApp, and generation of the `genes_set.tsv` used as input later, can be done by running [`00a_get_gene_set.sh`](./scripts/00a_get_gene_set.sh). Usage:

```
bash 00a_get_gene_set.sh /path/to/output/directory "https://panelapp.genomicsengland.co.uk/panels/402/download/34/"
```

This will store the gene panel and a `genes_set.tsv` file which looks like this:

```bash
ENSG00000078369
ENSG00000187730
ENSG00000198793
```

**IMPORTANT!** This script filters on gene names and assumes this information is contained in column 3.

### 2.  Get and process long-read RNA-seq data

This data can be downloaded using this [`00b_download_ENCODE_files.sh`](./scripts/00b_download_ENCODE_files.sh). It will calculate the total size of the download and allow a Y/N choice prior to downloading. Usage:

```
# first give the execute permission to the script:
chmod +x 00b_download_ENCODE_files.sh

# then run it:
00b_download_ENCODE_files.sh /path/to/directory/
```

To filter the data to only include the genes of interest (output from `genes_set.tsv` generated by `00a_get_gene_set.sh`) run [`00c_filter_ENCODE_long_read.sh`](./scripts/00c_filter_ENCODE_long_read.sh). Usage: 

```
bash 00c_filter_ENCODE_long_read.sh --directory /path/to/input/directory/ --output /path/to/output/directory/ --gene-file genes_set.tsv

```

### 3.  Get unique transcripts across samples

When retrieving novel transcripts through long-read RNA-seq, their IDs are generated in the order they are called. Consequently, IDs for the same transcript can vary among different samples. To ensure a consistent unique ID for all transcripts across samples (in `GTF` files), we execute run [`01a_process_ENCODE_transcript_structures.R`](./scripts/01a_process_ENCODE_transcript_structures.R) to get a set of unique transcripts.



## Contents

Within this repository you will find:
  
  | Directory | Description |
  | --------- | --------------------------------------------------------------------------- |
  | [raw_data](raw_data)| Data used for the analysis. Most will not be available due to size. |
  | [results](results) | Results from all analyses. |
  | [scripts](scripts) | Contains analysis scripts. Each script contains a one-line description and is also referenced in its corresponding `.Rmd`. |