# This script is used to filter the TX2P output
# 
# We ran all the transcripts through TX2P but will only include
# those expressed across all samples included and with a read cut-off

# Load libraries ----------------------------------------------------------

suppressPackageStartupMessages({
  library(tidyverse)
  library(readxl)
  library(here)
})

# Arguments ---------------------------------------------------------------

args <- list(
  path_to_TX2P = here::here("results", "TX2P_output", "EPI_Revised_main.xlsx"),
  path_to_expression = here::here("results", "expression_gene_set.rds")
)

# Load data ---------------------------------------------------------------

# TX2P output data
# This might need updating depending on the final output format we decide on for TX2P
# TX2P_count contains the number of times each transcript had a positive peptide detection
# TX2P_merged has all the data
TX2P_out <- list(
  TX2P_count = read_excel(args$path_to_TX2P, sheet = "Count_Found_in_mass_Datasets", col_names = c("data_set", "Transcript_ID", "occurences")),
  TX2P_merged = read_excel(args$path_to_TX2P, sheet = "merged_data")
  )

# This data is generated through 01b_process_ENCODE_expression.R
expression_gene_set <- readRDS(here::here("results", "expression_gene_set.rds"))


# Functions ---------------------------------------------------------------

#' filter transcripts
#'
#' @description  This function filter transcripts based on a set of parameters.
#'
#' @param data  `data.frame()` generated by 01b_process_ENCODE_expression.R
#' @param n_samples An integer defining the minimum number of samples the transcript was detcted in
#' @param n_reads An integer defining the minimum number of reads per sample
#' @param transcript_types character vector describing transcript types to include c("Known", "Genomic", "NNC", "ISM", "NIC")
#' @return `data.frame()` filtered by given parameters

filter_TX <- function(data, no_samples, no_reads, transcript_types) {
  TX_to_include <- 
    data %>% 
    dplyr::filter(transcript_novelty %in% transcript_types,
                  reads >= no_reads) %>% 
    dplyr::select(unique_id, sample) %>% 
    dplyr::group_by(unique_id) %>%
    dplyr::summarize(n_samples = n_distinct(sample)) %>%
    dplyr::filter(n_samples == no_samples) %>% 
    dplyr::pull(unique_id)
  
  filtered_data <- data %>%
    dplyr::filter(unique_id %in% TX_to_include)
  
  return(filtered_data)
}

# Example usage
filter_TX(data = expression_gene_set, no_samples = 9, no_reads = 1, transcript_types = c("NNC", "NIC"))







# Only include genes expressed across all samples
genes_to_include <- 
  expression_gene_set %>% 
  dplyr::select(annot_gene_id, sample) %>% 
  group_by(annot_gene_id) %>%
  summarize(n_samples = n_distinct(sample)) %>%
  dplyr::filter(n_samples == 9)

# Filter to only include transcripts of interest with novel ORFs from genes of interest
transcripts_to_include <- 
  expression_gene_set %>% 
  dplyr::select(annot_gene_id, unique_id, transcript_novelty, reads) %>% 
  dplyr::filter(annot_gene_id %in% genes_to_include$annot_gene_id,
                transcript_novelty %in% c("NNC", "NIC"), # only include transcripts with novel ORFs
                reads > 1) %>% # only include transcripts with at least 2 reads
  .$unique_id %>% 
  unique()


# Number of transcripts per transcript category
transcripts_to_include <- 
  expression_gene_set %>% 
  dplyr::select(annot_gene_id, unique_id, transcript_novelty, sample, reads) %>% 
  dplyr::filter(annot_gene_id %in% genes_to_include$annot_gene_id,
                transcript_novelty %in% c("NNC", "NIC"),
                reads > 1) %>% 
  dplyr::select(unique_id, sample) %>% 
  group_by(unique_id) %>%
  summarize(n_samples = n_distinct(sample)) %>% 
  dplyr::filter(n_samples > 1)




# filter data
final_transcript_set_MS <-
  mass_Spec_hits %>% 
  dplyr::filter(Transcript %in% transcripts_to_include$unique_id)

plot_MS_support <- 
  final_transcript_set_MS %>% 
  group_by(number_of_occurrences) %>%
  summarize(n_transcripts = n_distinct(Transcript)) %>% 
  
  ggplot(aes(x = number_of_occurrences,
             y = n_transcripts, 
             group = number_of_occurrences)) + 
  geom_col(aes(fill = factor(ifelse(number_of_occurrences > 0, "green", "lightgrey"))),
           colour = "black") +
  geom_text(aes(label = n_transcripts), 
            position = position_dodge(width=0.9), vjust= -0.6, size = 4) +
  labs(x = "No. mass spec data sets (total = 30)",
       y = "No. transcripts") +
  scale_fill_manual(values = c("green" = "#00B358", "lightgrey" = "lightgrey")) +
  theme_light() +
  guides(fill = "none") +
  theme(axis.title = element_text(size = 18),
        axis.text = element_text(size = 14),
        plot.margin = margin(0.1, 0.1, 0.1, 0.1, "cm"))

#### plot genes with MS support
transcript_set_MS_no_0 <- final_transcript_set_MS %>% dplyr::filter(number_of_occurrences > 0) 

gene_transcript <- 
  expression_gene_set %>% dplyr::select(annot_gene_name, RPM, unique_id) %>% 
  dplyr::filter(unique_id %in% transcripts_to_include$unique_id) %>% 
  aggregate(data = ., RPM ~ unique_id + annot_gene_name, FUN = mean)

final_MS_gene <- transcript_set_MS_no_0 %>% dplyr::left_join(., gene_transcript, by = c("Transcript" = "unique_id")) %>% 
  aggregate(data = ., RPM ~ annot_gene_name, FUN = sum) %>% 
  arrange(RPM)

plot_gene_novelty <- 
  final_MS_gene %>% 
  mutate(category = ifelse(RPM > 20, "Above 20 RPM", "Below or Equal 20 RPM")) %>% 
  ggplot(
    aes(x = factor(annot_gene_name, levels = annot_gene_name),
        y = RPM)) +
  geom_col(colour = "black",
           fill = "lightgrey") +
  labs(x = "Gene",
       y = "Expression of novel protein coding transcripts (RPM)") +
  coord_flip() +
  facet_wrap(~category, scales = "free", nrow = 1) +
  theme_light() +
  theme(legend.position = c(0.8, 0.8),
        legend.title = element_text(size = 10, face = "bold"),
        legend.background = element_rect(fill = alpha("white", 0.0)),
        axis.title = element_text(size = 14),
        axis.text.x = element_text(size = 10),
        axis.text.y = element_text(size = 7))

########### How much does this novel protein coding transcript correspond to in relative terms of expression ##########
genes_with_novel_prot <- 
  expression_gene_set %>% 
  dplyr::filter(unique_id %in% transcript_set_MS_no_0$Transcript) %>% 
  dplyr::select(annot_gene_name) %>% 
  unique()

total_expression_per_sample <-
  expression_gene_set %>% 
  dplyr::filter(annot_gene_name %in% genes_with_novel_prot$annot_gene_name) %>% 
  aggregate(data = ., RPM ~ annot_gene_name + sample, FUN = sum)


novel_expression <-
  expression_gene_set %>% 
  dplyr::filter(unique_id %in% transcript_set_MS_no_0$Transcript) %>% 
  aggregate(data = ., RPM ~ annot_gene_name + sample, FUN = sum)

combined <- 
  total_expression_per_sample %>% 
  dplyr::left_join(., novel_expression, by = join_by(annot_gene_name, sample)) %>% 
  replace(., is.na(.), 0) %>% 
  dplyr::mutate(novel_expression_perc = (RPM.y / RPM.x) *100)

top_genes <- 
  combined %>% 
  aggregate(data = ., novel_expression_perc ~ annot_gene_name, FUN = mean) %>% 
  arrange(desc(novel_expression_perc)) %>% 
  slice_max(novel_expression_perc, n = 20)



plot_novel_expression_all <-
  combined %>% 
  ggplot(
    aes(x = reorder(annot_gene_name, -novel_expression_perc),
        y = novel_expression_perc,
        group = annot_gene_name)) +
  geom_boxplot(colour = "black", fill = "lightgrey", show.legend = F, outlier.shape = NA) +
  labs(x = "Gene",
       y = "Relative expression of novel protein coding transcripts (%)") +
  theme_light() +
  theme(axis.title = element_text(size = 14),
        axis.text.y = element_text(size = 10),
        axis.text.x = element_text(face = "bold", size = 8, angle = 45, hjust = 1)
)

plot_novel_expression_top <-
  combined %>% 
  dplyr::filter(annot_gene_name %in% top_genes$annot_gene_name) %>% 
  ggplot(
    aes(x = reorder(annot_gene_name, -novel_expression_perc),
        y = novel_expression_perc,
        group = annot_gene_name)) +
  geom_boxplot(colour = "black", fill = "lightgrey", show.legend = F, outlier.shape = NA) +
  geom_point() +
  labs(x = "Gene",
       y = "Relative expression of novel protein coding transcripts (%)") +
  theme_light() +
  theme(axis.title = element_text(size = 14),
        axis.text.y = element_text(size = 10),
        axis.text.x = element_text(face = "bold", size = 12, angle = 45, hjust = 1)
  )

# Save data ---------------------------------------------------------------

ggsave(
  plot = plot_MS_support, 
  filename = "plot_MS_support.png", 
  path = here::here("results", "plots"), 
  width = 8, 
  height = 6, 
  dpi = 600, 
  bg = "white"
)

ggsave(
  plot = plot_gene_novelty, 
  filename = "plot_gene_novelty.png", 
  path = here::here("results", "plots"), 
  width = 16, 
  height = 8, 
  dpi = 600, 
  bg = "white"
)

ggsave(
  plot = plot_novel_expression_all, 
  filename = "plot_novel_expression_all.png", 
  path = here::here("results", "plots"), 
  width = 16, 
  height = 8, 
  dpi = 600, 
  bg = "white"
)

ggsave(
  plot = plot_novel_expression_top, 
  filename = "plot_novel_expression_top.png", 
  path = here::here("results", "plots"), 
  width = 12, 
  height = 8, 
  dpi = 600, 
  bg = "white"
)